File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
    var answer = 0;
    const set = new Set();
    
    function bfs(i) {
        const queue = [];
        queue.push(i);
        set.add(i);
        
        let head = 0;
        while(head < queue.length) {
            const item = queue[head++];
            for(let i = 0; i < n; i++) {
                if(computers[item][i] === 1 && !set.has(i) && item !== i) {
                    queue.push(i);
                    set.add(i);
                }
            }
        }
    }
    
    for(let i = 0; i < n; i++) {
        if(set.has(i)) continue;
        bfs(i);
        answer++;
    }
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    HashSet<Integer> set =  new HashSet<>();
    
    public void bfs(int n, int[][] computers) {
        Deque<Integer> queue = new ArrayDeque<>();
        queue.push(n);
        set.add(n);
        
        while(queue.size() > 0) {
            int item = queue.poll();
            for(int i = 0; i < computers[item].length; i++) {
                if(set.contains(i) || computers[item][i] == 0 || i == n) continue;
                queue.push(i);
                set.add(i);
            }
        }
    }
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        for(int i = 0; i < n; i++) {
            if(set.contains(i)) continue;
            bfs(i, computers);
            answer++;
        }
        
        return answer;
    }
}